File: Tests/TamaCoreTests/OllamaReaderTests.swift
import XCTest
import TamaCore

final class OllamaReaderTests: XCTestCase {
    private let fm = FileManager.default
    private var utc: Calendar { var c = Calendar(identifier: .gregorian); c.timeZone = TimeZone(identifier: "UTC")!; return c }

    private func fixture(_ lines: [String]) throws -> URL {
        let dir = fm.temporaryDirectory.appendingPathComponent("ollama-\(UUID().uuidString)/logs")
        try fm.createDirectory(at: dir, withIntermediateDirectories: true)
        let url = dir.appendingPathComponent("server.log")
        try lines.joined(separator: "\n").write(to: url, atomically: true, encoding: .utf8)
        return url
    }

    // Real-format line builders.
    private func start(_ tag: String, port: Int = 100) -> String {
        #"time=2026-06-22T18:36:56.556+07:00 level=INFO source=client.go:367 msg="starting mlx runner subprocess" model=\#(tag) port=\#(port)"#
    }
    private func stop(pid: Int = 999) -> String {
        #"time=2026-06-22T19:29:32.938+07:00 level=INFO source=client.go:126 msg="stopping mlx runner subprocess" pid=\#(pid)"#
    }
    private func prompt(_ win: Int, _ tok: Int) -> String {
        "slot update_slots: id  0 | task 5 | new prompt, n_ctx_slot = \(win), n_keep = 4, task.n_tokens = \(tok)"
    }
    private let processing = "slot launch_slot_: id  0 | task 5 | processing task, is_child = 0"
    private let idle = "srv  update_slots: all slots are idle"
    private func timing(_ tps: Double) -> String {
        "slot print_timing: id  0 | task 5 | n_decoded =    100, tg =  \(tps) t/s"
    }
    private func gin(_ time: String, _ dur: String, _ endpoint: String, status: Int = 200) -> String {
        "[GIN] 2026/06/22 - \(time) | \(status) |   \(dur) |       127.0.0.1 | POST     \"\(endpoint)\""
    }

    func test_lists_models_most_recent_first_with_attribution() throws {
        let url = try fixture([
            start("qwen3.6:27b-mlx"), prompt(32768, 2000), processing, timing(50.0),
            gin("19:00:00", "5.0s", "/api/generate"), idle, stop(),
            start("gemma4:12b-mlx"), prompt(8192, 500), processing, timing(80.0),
            gin("19:30:00", "2.0s", "/api/chat"), gin("19:30:05", "200ms", "/api/chat"),
        ])
        let models = OllamaReader(logURL: url, calendar: utc).read()?.models
        XCTAssertEqual(models?.count, 2)

        let g = models?.first
        XCTAssertEqual(g?.model, "gemma4:12b-mlx")
        XCTAssertEqual(g?.current, true)
        XCTAssertEqual(g?.busy, true)                       // processing, no idle after
        XCTAssertEqual(g?.contextWindow, 8192)
        XCTAssertEqual(g?.contextTokens, 500)
        XCTAssertEqual(g?.tokensPerSecond, 80.0)
        XCTAssertEqual(g?.lastLatency ?? 0, 0.2, accuracy: 0.001)   // 200ms
        XCTAssertEqual(g?.kind, .chat)
        XCTAssertEqual(g?.requestCount, 2)

        let q = models?.last
        XCTAssertEqual(q?.model, "qwen3.6:27b-mlx")
        XCTAssertEqual(q?.current, false)
        XCTAssertEqual(q?.busy, false)                      // idle after
        XCTAssertEqual(q?.contextWindow, 32768)
        XCTAssertEqual(q?.tokensPerSecond, 50.0)
        XCTAssertEqual(q?.lastLatency ?? 0, 5.0, accuracy: 0.001)
        XCTAssertEqual(q?.requestCount, 1)
    }

    func test_embedding_endpoint_is_embed_kind() throws {
        let url = try fixture([start("bge-m3"), gin("19:00:00", "120ms", "/api/embeddings")])
        XCTAssertEqual(OllamaReader(logURL: url, calendar: utc).read()?.models.first?.kind, .embed)
    }

    func test_same_model_loaded_twice_merges() throws {
        let url = try fixture([
            start("gemma4:12b-mlx"), gin("19:00:00", "1.0s", "/api/chat"), idle, stop(),
            start("gemma4:12b-mlx"), gin("19:10:00", "1.0s", "/api/chat"),
        ])
        let models = OllamaReader(logURL: url, calendar: utc).read()?.models
        XCTAssertEqual(models?.count, 1)
        XCTAssertEqual(models?.first?.requestCount, 2)
    }

    func test_non_inference_endpoints_do_not_count() throws {
        let url = try fixture([start("gemma4:12b-mlx"), gin("19:00:00", "10ms", "/api/tags"),
                               gin("19:00:01", "5ms", "/api/show")])
        let m = OllamaReader(logURL: url, calendar: utc).read()?.models.first
        XCTAssertEqual(m?.requestCount, 0)
        XCTAssertEqual(m?.kind, .unknown)
    }

    func test_last_activity_is_log_mtime() throws {
        let url = try fixture([start("gemma4:12b-mlx"), idle])
        let stamp = Date(timeIntervalSince1970: 1_700_000_000)
        try fm.setAttributes([.modificationDate: stamp], ofItemAtPath: url.path)
        let last = try XCTUnwrap(OllamaReader(logURL: url, calendar: utc).read()?.lastActivity)
        XCTAssertEqual(last.timeIntervalSince1970, stamp.timeIntervalSince1970, accuracy: 1)
    }

    func test_running_with_no_model_loads_yields_empty_models() throws {
        let url = try fixture(["srv  update_slots: all slots are idle", "some unrelated line"])
        let status = OllamaReader(logURL: url, calendar: utc).read()
        XCTAssertNotNil(status)
        XCTAssertEqual(status?.models.isEmpty, true)
    }

    func test_missing_log_is_nil() {
        let missing = fm.temporaryDirectory.appendingPathComponent("nope-\(UUID().uuidString)/server.log")
        XCTAssertNil(OllamaReader(logURL: missing, calendar: utc).read())
    }

    func test_empty_log_is_nil() throws {
        XCTAssertNil(OllamaReader(logURL: try fixture([]), calendar: utc).read())
    }
}

File: windows/Tama.Core.Tests/OllamaReaderTests.cs
using Tama.Core;

namespace Tama.Core.Tests;

[TestClass]
public sealed class OllamaReaderTests
{
    private string _root = null!;

    [TestInitialize]
    public void SetUp()
    {
        _root = Path.Combine(Path.GetTempPath(), "ollama-" + Guid.NewGuid());
        Directory.CreateDirectory(_root);
    }

    [TestCleanup]
    public void TearDown() => Directory.Delete(_root, recursive: true);

    private string Fixture(params string[] lines)
    {
        var p = Path.Combine(_root, "server.log");
        File.WriteAllText(p, string.Join("\n", lines));
        return p;
    }

    private static string Start(string tag) =>
        $"time=2026-06-22T18:36:56.556+07:00 level=INFO source=client.go:367 msg=\"starting mlx runner subprocess\" model={tag} port=100";
    private const string Stop =
        "time=2026-06-22T19:29:32.938+07:00 level=INFO source=client.go:126 msg=\"stopping mlx runner subprocess\" pid=999";
    private static string Prompt(int win, int tok) =>
        $"slot update_slots: id  0 | task 5 | new prompt, n_ctx_slot = {win}, n_keep = 4, task.n_tokens = {tok}";
    private const string Processing = "slot launch_slot_: id  0 | task 5 | processing task, is_child = 0";
    private const string Idle = "srv  update_slots: all slots are idle";
    private static string Timing(double tps) =>
        $"slot print_timing: id  0 | task 5 | n_decoded =    100, tg =  {tps} t/s";
    private static string Gin(string time, string dur, string endpoint) =>
        $"[GIN] 2026/06/22 - {time} | 200 |   {dur} |       127.0.0.1 | POST     \"{endpoint}\"";

    private static OllamaReader Reader(string path) => new(path, TimeZoneInfo.Utc);

    [TestMethod]
    public void Lists_models_current_first_with_attribution()
    {
        var p = Fixture(
            Start("qwen3.6:27b-mlx"), Prompt(32768, 2000), Processing, Timing(50.0),
            Gin("19:00:00", "5.0s", "/api/generate"), Idle, Stop,
            Start("gemma4:12b-mlx"), Prompt(8192, 500), Processing, Timing(80.0),
            Gin("19:30:00", "2.0s", "/api/chat"), Gin("19:30:05", "200ms", "/api/chat"));
        var models = Reader(p).Read()!.Models;
        Assert.AreEqual(2, models.Count);

        var g = models[0];
        Assert.AreEqual("gemma4:12b-mlx", g.Model);
        Assert.IsTrue(g.Current);
        Assert.IsTrue(g.Busy);                        // processing, no idle after
        Assert.AreEqual(8192, g.ContextWindow);
        Assert.AreEqual(500, g.ContextTokens);
        Assert.AreEqual(80.0, g.TokensPerSecond);
        Assert.AreEqual(0.2, g.LastLatencySeconds!.Value, 0.001);   // 200ms
        Assert.AreEqual(OllamaRequestKind.Chat, g.Kind);
        Assert.AreEqual(2, g.RequestCount);
        Assert.AreEqual(new DateTimeOffset(2026, 6, 22, 19, 30, 5, TimeSpan.Zero), g.LastActivity);

        var q = models[1];
        Assert.AreEqual("qwen3.6:27b-mlx", q.Model);
        Assert.IsFalse(q.Current);
        Assert.IsFalse(q.Busy);                       // idle after
        Assert.AreEqual(32768, q.ContextWindow);
    }

    [TestMethod]
    public void Non_inference_endpoints_do_not_count()
    {
        var p = Fixture(Start("m:1"),
            Gin("19:00:00", "1ms", "/api/tags"), Gin("19:00:01", "1ms", "/api/ps"));
        var m = Reader(p).Read()!.Models[0];
        Assert.AreEqual(0, m.RequestCount);
        Assert.AreEqual(OllamaRequestKind.Unknown, m.Kind);
    }

    [TestMethod]
    public void Embed_majority_flips_kind()
    {
        var p = Fixture(Start("m:1"),
            Gin("19:00:00", "1ms", "/api/embed"), Gin("19:00:01", "1ms", "/api/embed"),
            Gin("19:00:02", "1ms", "/api/chat"));
        Assert.AreEqual(OllamaRequestKind.Embed, Reader(p).Read()!.Models[0].Kind);
    }

    [TestMethod]
    public void Lines_before_any_start_are_unattributed()
    {
        var p = Fixture(Prompt(1000, 10), Gin("19:00:00", "1s", "/api/chat"), Start("m:1"));
        var m = Reader(p).Read()!.Models[0];
        Assert.AreEqual(0, m.RequestCount);
        Assert.IsNull(m.ContextWindow);
    }

    [TestMethod]
    public void Reload_of_same_model_merges_into_one_entry()
    {
        var p = Fixture(
            Start("m:1"), Gin("19:00:00", "1s", "/api/chat"), Stop,
            Start("m:1"), Gin("19:10:00", "1s", "/api/chat"));
        var models = Reader(p).Read()!.Models;
        Assert.AreEqual(1, models.Count);
        Assert.AreEqual(2, models[0].RequestCount);
    }

    [TestMethod]
    public void Go_durations_parse_all_suffixes()
    {
        var p = Fixture(Start("m:1"),
            Gin("19:00:00", "23.917µs", "/api/chat"));
        Assert.AreEqual(23.917e-6, Reader(p).Read()!.Models[0].LastLatencySeconds!.Value, 1e-9);
    }

    [TestMethod]
    public void Missing_or_empty_log_returns_null()
    {
        Assert.IsNull(Reader(Path.Combine(_root, "nope.log")).Read());
        Assert.IsNull(Reader(Fixture()).Read());
    }
}
